src/access.lua

```diff
@@ -58,45 +58,53 @@ function _M.run(conf)
             end
 
             -- Get user info
-            -- TODO: don't know if makes sense to do everytime this call to check the validity of the token
-            --  or cache the result in a Cookie that expires early
-            local httpc = http:new()
-            local res, err = httpc:request_uri(conf.user_url, {
-                method = "GET",
-                ssl_verify = false,
-                headers = {
-                  ["Authorization"] = "Bearer " .. access_token,
-                }
-            })
-
-            if res then
-                -- redirect to auth if user result is invalid not 200
-                if res.status ~= 200 then
-                    return redirect_to_auth( conf, callback_url )
-                end
+            if not ngx.var.cookie_EOAuthUserInfo then
+                local httpc = http:new()
+                local res, err = httpc:request_uri(conf.user_url, {
+                    method = "GET",
+                    ssl_verify = false,
+                    headers = {
+                      ["Authorization"] = "Bearer " .. access_token,
+                    }
+                })
+
+                if res then
+                    -- redirect to auth if user result is invalid not 200
+                    if res.status ~= 200 then
+                        return redirect_to_auth( conf, callback_url )
+                    end
 
-                local json = cjson.decode(res.body)
+                    local json = cjson.decode(res.body)
 
-                if conf.hosted_domain ~= "" and conf.email_key ~= "" then
-                    if not pl_stringx.endswith(json[conf.email_key], conf.hosted_domain) then
-                        ngx.status = 401
-                        ngx.say("Hosted domain is not matching")
-                        ngx.exit(ngx.HTTP_OK)
-                        return
+                    if conf.hosted_domain ~= "" and conf.email_key ~= "" then
+                        if not pl_stringx.endswith(json[conf.email_key], conf.hosted_domain) then
+                            ngx.status = 401
+                            ngx.say("Hosted domain is not matching")
+                            ngx.exit(ngx.HTTP_OK)
+                            return
+                        end
+                    end
+
+                    for i, key in ipairs(conf.user_keys) do
+                        ngx.header["X-Oauth-".. key] = json[key]
+                    end
+                    ngx.header["X-Oauth-Token"] = access_token
+
+                    if type(ngx.header["Set-Cookie"]) == "table" then
```

Comment:
By the way, I think the `path` could be more specific using the `request_uri`; but I'd like to have some tests first.